docs/structured.md

```diff
@@ -0,0 +1,212 @@
+##About##
+
+This document defines the structured data layer that Cockroach exposes on top of
+its distributed key:value datastore.
+
+##Overview##
+
+The structured data layer is an abstraction over the internal key:value store
+that provides familiar data organization tools. The structured data layer
+implements a tabular model of data storage which somewhat resembles a relational
+database at a high level. The terms namespace, table, column, and index/key
+collectively define the structured data abstraction, each roughly akin to
+database, table, column, and index/key in a relational database. The structured
+data layer introduces concepts of data storage that many
+database users are familiar with, and greatly increases the range of use-cases
+that Cockroach is able to ergonomically service.
+
+##Goals##
+
+Add support for the following entities: *namespaces, tables, columns, and
+indexes*. These notions are chosen for their similarity to *databases, tables,
+columns, and indexes* in relational databases, but they are not intended to be
+implementations thereof. Each entity will support:
+
+- Creation: of a namespace, table, column in O(1).
+
+- Renaming: in O(1) Most configuration changes in O(1): A change in permissions
+  for instance, will be in O(1) time. Adding a new index will take O(n) time,
+where n is the number of rows in the table.
+
+- Deletion: in O(n) where n is the number of subordinate entities (tables for a
+  namespace, rows for a table).
+
+The client API will be changed to use tabular based addressing instead of
+global-key addressing. Non-primary-keyed queries will be supported through the use of secondary indexes.
+
+##Non-Goals##
+
+For now we don't intend on developing the following:
+
+- Add support for triggers, stored-procedures, and integrity constraints
+
+- Add the notion of column types beyond *a sequence of bytes*. Add support for
+  unbounded sub-collections of columns.
+
+- Add support for non-tabular storage. This might be supported in the future.
+
+- Add support for column locality-groups to improve data layout for performance.
+
+##Design##
+
+We support high level entities like Namespace, Table, Index, and Column.
+{Namespace,Table}Descriptor stores the entity metadata for {Namespace,Table}.
+
+```proto
+NamespaceDescriptor {
+  NamespaceID = … ,
+  Permissions = …,
+  ...}
```

Comment:
put this ellipsis on the above line so that the closing brace is correctly aligned